src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs

```diff
@@ -81,7 +135,106 @@ public T DeserializePayload<T>(Message message)
     {
         var versionedMessage = message as VersionedMessage;
         var payloadSerializer = GetPayloadSerializer(versionedMessage?.Version);
-        return Deserialize<T>(payloadSerializer, message.Payload);
+
+        if (!DisableFastJson)
+        {
+            var messageWithRawMessage = message as MessageWithRawMessage;
+            var rawMessage = messageWithRawMessage.RawMessage;
+
+            if (payloadSerializer == s_payloadSerializer2)
+            {
+                // PERF: Fast path is compatibile only with protocol versions that use serializer_2,
+                // and this is faster than deserializing via deserializer_2.
+                var messageWithPayload = JsonConvert.DeserializeObject<PayloadedMessage<T>>(rawMessage, s_jsonSettings7);
+                return messageWithPayload.Payload;
+            }
+            else
+            {
+                // PERF: When payloadSerializer1 was resolved we need to deserialize JToken, and then deserialize that.
+                // This is still better than deserializing the JToken in DeserializeMessage because here we know that the payload
+                // will actually be used.
+                return Deserialize<T>(payloadSerializer, Deserialize<VersionedMessage>(rawMessage).Payload);
+            }
+        }
+        else
+        {
+            return Deserialize<T>(payloadSerializer, message.Payload);
+        }
+    }
+
+    private bool FastHeaderParse(string rawMessage, out int version, out string messageType)
```

Comment:
General not mandatory info, if you start to check from the highest possible index all lower access will be optimized without range check by jitter.

https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgMwAEOZAwmQN5FmMXmwJkCyAFKwNoC6ZCAEo6DJuNZkATgFMUAVwA2bALxkADAG4x4xjt2yFysgGo1EHtj7bCupoaVszgnrmv7xD484vr3tuzIPJkwAdmk5RxtxAF99fVIySXZsbhgEfkERegC7SS9VDWjA4MYC03MeP2K7cp9Xf0CIoydKqxrdUopwgo64whigA==

Reply:
I don't know the highest index until I find the last message type `"`, but I don't know where to start looking for it, until after I know how long the version number is. 

Or is there some trick you can think about?

Reply:
I mean you check for 2 and after for 10...so if 10 is there 2 is there...you can do it until the "known" length. 
But I think it's too much and risky maybe to skip a pair of test. 

Take it as an interesting info, it's ok as-is.

Reply:
Oh I've missed the check for the malformed messages...so I think that now it's optimized until the 31 index access.